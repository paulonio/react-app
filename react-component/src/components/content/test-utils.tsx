import React, { FC, ReactElement, useReducer } from 'react';
import { initialState, mainReducer, SearchContext } from 'context/context';

export const StateProvider: FC<{ children: React.ReactNode }> = ({ children }) => {
  const [state, dispatch] = useReducer(mainReducer, initialState);
  return <SearchContext.Provider value={{ state, dispatch }}>{children}</SearchContext.Provider>;
};
