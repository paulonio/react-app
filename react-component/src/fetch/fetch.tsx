export const fetchData = async (page: string) => {
  const baseURL = `https://rickandmortyapi.com/api/character/?page=${page}`;
  const response = await (await fetch(baseURL)).json();
  return response;
};

export const fetchByCharacterName = async (query: string, page: string) => {
  try {
    const baseURL = `https://rickandmortyapi.com/api/character/?name=${query}&page=${page}`;
    const response = await fetch(baseURL);
    return response.status === 200 ? await response.json() : null;
  } catch (err) {}
};
