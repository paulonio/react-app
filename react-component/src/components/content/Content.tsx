import React, { useContext, useEffect, useState } from 'react';

import { fetchByCharacterName, fetchData } from 'fetch/fetch';
import { SearchContext } from 'context/context';
import { Action } from 'context/reducers';

import Cards from 'components/cards/Cards';
import ContentHeader from 'components/content-header/ContentHeader';
import Loader from 'components/loader/Loader';
import Modal from 'components/modal/Modal';
import CustomSelect from './utils/CustomSelect';
import ContentButtons from './content-buttons/ContentButtons';

import classes from './content.module.scss';

export interface IDataRick {
  name?: string;
  created?: string;
  episode?: string[];
  gender?: string;
  id?: number;
  image?: string;
  location?: {
    name: string;
    url: string;
  };
  origin?: {
    name: string;
    url: string;
  };
  species?: string;
  status?: string;
  type?: string;
  url?: string;
}

export type OptionsType = { value: string; label: string };

const sortOptions: OptionsType[] = [
  { value: 'by-creation', label: 'By creation' },
  { value: 'by-name', label: 'By name' },
  { value: 'by-species', label: 'By species' },
];

const paginationOptions: OptionsType[] = [
  { value: '12', label: '12 per page' },
  { value: '16', label: '16 per page' },
  { value: '20', label: '20 per page' },
];

const Content = () => {
  const [modalIsActive, setModalActive] = useState<boolean>(false);
  const [modalData, setModalData] = useState<IDataRick>({});
  const [isLoading, setLoading] = useState<boolean>(false);
  const [totalPages, setTotalPages] = useState<number>(1);
  const { state, dispatch } = useContext(SearchContext);
  const page = state.pageNumber || '1';

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const action: Action = { type: 'UPDATE_SEARCH_VALUE', value: e.target.value };
    dispatch(action);
  };

  const handleSortChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const action: Action = { type: 'UPDATE_SORT_OPTION', option: e.target.value };
    dispatch(action);
  };

  const handlePaginationChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const action: Action = { type: 'UPDATE_ITEMS_PER_PAGE', value: e.target.value };
    dispatch(action);
  };

  const sortData = (data: IDataRick[]) => {
    const option = state.sortOption;
    if (option === 'by-name') {
      return data.sort((a, b) => a.name!.localeCompare(b.name!) - b.name!.localeCompare(a.name!));
    }
    if (option === 'by-creation') {
      return data.sort((a, b) => Date.parse(a.created!) - Date.parse(b.created!));
    }
    if (option === 'by-species') {
      return data.sort(
        (a, b) => a.species!.localeCompare(b.species!) - b.species!.localeCompare(a.species!)
      );
    }
    return data;
  };

  const sliceResults = (data: IDataRick[]) => {
    const maxCardsOnPage = Number(state.paginationValue);
    return data.slice(0, maxCardsOnPage);
  };

  const getDataOnMount = async () => {
    setLoading(true);
    const response = await fetchData(page);
    if (response) {
      setTotalPages(response.info.pages);
      const result: IDataRick[] = response.results;
      const correctNumber = sliceResults(result);
      const sortedData = sortData(correctNumber);
      const action: Action = { type: 'GET_RICK_DATA', data: sortedData };
      dispatch(action);
    }
    setLoading(false);
  };

  const handleSearch = async (query: string, page?: string) => {
    setLoading(true);
    const currentPage = state.pageNumber;
    const response = await fetchByCharacterName(query, page || currentPage);
    if (response) {
      setTotalPages(response.info.pages);
      const result: IDataRick[] = response.results;
      const correctNumber = result && sliceResults(result);
      const sortedData = correctNumber && sortData(correctNumber);
      const action: Action = { type: 'GET_RICK_DATA', data: sortedData };
      dispatch(action);
    } else {
      const action: Action = { type: 'GET_RICK_DATA', data: [] };
      dispatch(action);
    }
    setLoading(false);
  };

  const handleKeyboardEvent = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.code === 'Enter') {
      const actionPage: Action = { type: 'UPDATE_PAGE_NUMBER', page: '1' };
      dispatch(actionPage);
      handleSearch(state.searchValue, '1');
    }
  };

  const handleCardClick = (e: React.MouseEvent<HTMLElement>) => {
    const target = e.currentTarget as HTMLElement;
    const idx = target.dataset.id && target.dataset.id;
    const [targetData] = state.rickData!.filter((item) => idx === String(item.id));
    setModalData(targetData);
  };

  useEffect(() => {
    const value = state.searchValue;
    value ? handleSearch(value) : getDataOnMount();
  }, [state.sortOption, state.paginationValue, state.pageNumber]);

  return (
    <section className={classes.home__content}>
      <ContentHeader
        inputValue={state.searchValue}
        handleChange={handleChange}
        handleKeyboardEvent={handleKeyboardEvent}
        handleSearch={handleSearch}
      />
      <div className={classes.select__container}>
        <CustomSelect handler={handleSortChange} options={sortOptions} />
        <CustomSelect handler={handlePaginationChange} options={paginationOptions} />
      </div>
      {state.rickData && (
        <Cards
          data={state.rickData}
          modalSetActive={setModalActive}
          handleCardClick={handleCardClick}
        />
      )}
      {!state.rickData && <div className={classes.error__message}>Sorry, data not found.</div>}
      {modalData.name && (
        <Modal isActive={modalIsActive} setActive={setModalActive} targetData={modalData} />
      )}
      {isLoading && <Loader />}
      {!!state.rickData && <ContentButtons totalPages={totalPages} />}
    </section>
  );
};

export default Content;
