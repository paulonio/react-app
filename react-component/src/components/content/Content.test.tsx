import React from 'react';
import { waitFor } from '@testing-library/react';
import { StateProvider } from './test-utils';
import renderer, { act } from 'react-test-renderer';
import Content from './Content';
import { fetchData } from '../../fetch/fetch';

const MOCK_DATA = {
  info: {
    count: 29,
    pages: 2,
    next: 'https://rickandmortyapi.com/api/character/?page=2&name=rick&status=alive',
    prev: null,
  },
  results: [
    {
      id: 1,
      name: 'Mocked name',
      status: 'Alive',
      species: 'Human',
      type: '',
      gender: 'Male',
      origin: {
        name: 'Earth',
        url: 'https://rickandmortyapi.com/api/location/1',
      },
      location: {
        name: 'Earth',
        url: 'https://rickandmortyapi.com/api/location/20',
      },
      image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
      episode: [
        'https://rickandmortyapi.com/api/episode/1',
        'https://rickandmortyapi.com/api/episode/2',
      ],
      url: 'https://rickandmortyapi.com/api/character/1',
      created: '2017-11-04T18:48:46.250Z',
    },
  ],
};

jest.mock('components/fetch/fetch', () => ({
  fetchData: jest.fn(),
  fetchByCharacterName: jest.fn(),
}));

jest.mock('components/cards/Cards', () => `Cards`);
jest.mock('components/content-header/ContentHeader', () => `ContentHeader`);
jest.mock('./utils/CustomSelect', () => `CustomSelect`);
jest.mock('components/loader/Loader', () => `Loader`);
jest.mock('components/modal/Modal', () => `Modal`);
jest.mock('./content-buttons/ContentButtons', () => `ContentButtons`);

describe('Content', () => {
  const mockFetchData = fetchData as jest.MockedFunction<typeof fetchData>;

  afterEach(() => {
    jest.clearAllMocks();
    jest.restoreAllMocks();
    jest.resetAllMocks();
  });

  it('renders content component on mount', async () => {
    const tree = renderer.create(
      <StateProvider>
        <Content />
      </StateProvider>
    );

    expect(tree).toMatchSnapshot();

    await act(async () => {
      await waitFor(() => expect(mockFetchData).toHaveBeenCalledTimes(1));
    });
  });

  it('renders content component after data loaded', async () => {
    mockFetchData.mockResolvedValue(MOCK_DATA);

    const tree = renderer.create(
      <StateProvider>
        <Content />
      </StateProvider>
    );

    await act(async () => {
      await waitFor(() => {
        expect(mockFetchData).toHaveBeenCalledTimes(1);
      });
    });

    expect(tree).toMatchSnapshot();
  });

  it('renders content component after fetch rejected', async () => {
    mockFetchData.mockResolvedValue(null);

    const tree = renderer.create(
      <StateProvider>
        <Content />
      </StateProvider>
    );

    await act(async () => {
      await waitFor(() => expect(mockFetchData).toHaveBeenCalledTimes(1));
    });
    expect(tree).toMatchSnapshot();
  });
});
