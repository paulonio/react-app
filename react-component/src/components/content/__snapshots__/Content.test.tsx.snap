// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content render content component with data 1`] = `null`;

exports[`Content renders content component after data loaded 1`] = `
<section
  className="home__content"
>
  <ContentHeader
    handleChange={[Function]}
    handleKeyboardEvent={[Function]}
    handleSearch={[Function]}
    inputValue=""
  />
  <div
    className="select__container"
  >
    <CustomSelect
      handler={[Function]}
      options={
        Array [
          Object {
            "label": "By creation",
            "value": "by-creation",
          },
          Object {
            "label": "By name",
            "value": "by-name",
          },
          Object {
            "label": "By species",
            "value": "by-species",
          },
        ]
      }
    />
    <CustomSelect
      handler={[Function]}
      options={
        Array [
          Object {
            "label": "12 per page",
            "value": "12",
          },
          Object {
            "label": "16 per page",
            "value": "16",
          },
          Object {
            "label": "20 per page",
            "value": "20",
          },
        ]
      }
    />
  </div>
  <Cards
    data={
      Array [
        Object {
          "created": "2017-11-04T18:48:46.250Z",
          "episode": Array [
            "https://rickandmortyapi.com/api/episode/1",
            "https://rickandmortyapi.com/api/episode/2",
          ],
          "gender": "Male",
          "id": 1,
          "image": "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
          "location": Object {
            "name": "Earth",
            "url": "https://rickandmortyapi.com/api/location/20",
          },
          "name": "Mocked name",
          "origin": Object {
            "name": "Earth",
            "url": "https://rickandmortyapi.com/api/location/1",
          },
          "species": "Human",
          "status": "Alive",
          "type": "",
          "url": "https://rickandmortyapi.com/api/character/1",
        },
      ]
    }
    handleCardClick={[Function]}
    modalSetActive={[Function]}
  />
  <ContentButtons
    totalPages={2}
  />
</section>
`;

exports[`Content renders content component after fetch rejected 1`] = `
<section
  className="home__content"
>
  <ContentHeader
    handleChange={[Function]}
    handleKeyboardEvent={[Function]}
    handleSearch={[Function]}
    inputValue=""
  />
  <div
    className="select__container"
  >
    <CustomSelect
      handler={[Function]}
      options={
        Array [
          Object {
            "label": "By creation",
            "value": "by-creation",
          },
          Object {
            "label": "By name",
            "value": "by-name",
          },
          Object {
            "label": "By species",
            "value": "by-species",
          },
        ]
      }
    />
    <CustomSelect
      handler={[Function]}
      options={
        Array [
          Object {
            "label": "12 per page",
            "value": "12",
          },
          Object {
            "label": "16 per page",
            "value": "16",
          },
          Object {
            "label": "20 per page",
            "value": "20",
          },
        ]
      }
    />
  </div>
  <div
    className="error__message"
  >
    Sorry, data not found. 
    state: null
  </div>
</section>
`;

exports[`Content renders content component on mount 1`] = `
<section
  className="home__content"
>
  <ContentHeader
    handleChange={[Function]}
    handleKeyboardEvent={[Function]}
    handleSearch={[Function]}
    inputValue=""
  />
  <div
    className="select__container"
  >
    <CustomSelect
      handler={[Function]}
      options={
        Array [
          Object {
            "label": "By creation",
            "value": "by-creation",
          },
          Object {
            "label": "By name",
            "value": "by-name",
          },
          Object {
            "label": "By species",
            "value": "by-species",
          },
        ]
      }
    />
    <CustomSelect
      handler={[Function]}
      options={
        Array [
          Object {
            "label": "12 per page",
            "value": "12",
          },
          Object {
            "label": "16 per page",
            "value": "16",
          },
          Object {
            "label": "20 per page",
            "value": "20",
          },
        ]
      }
    />
  </div>
  <div
    className="error__message"
  >
    Sorry, data not found. 
    state: null
  </div>
</section>
`;
